src/indexBy.ts

```diff
@@ -33,33 +33,35 @@ export function indexBy<T>(
  * @indexed
  * @category Array
  */
-export function indexBy<T>(
-  fn: (item: T) => any
-): (array: ReadonlyArray<T>) => Record<string, T>;
+export function indexBy<K extends PropertyKey, T>(
+  fn: (item: T) => K
+): (array: ReadonlyArray<T>) => Record<K, T>;
 
 export function indexBy() {
   return purry(_indexBy(false), arguments);
 }
 
 const _indexBy =
   (indexed: boolean) =>
-  <T>(array: Array<T>, fn: PredIndexedOptional<T, any>) => {
-    return array.reduce<Record<string, T>>((ret, item, index) => {
-      const value = indexed ? fn(item, index, array) : fn(item);
-      const key = String(value);
+  <K extends PropertyKey, T>(
+    array: Array<T>,
+    fn: PredIndexedOptional<T, K>
+  ) => {
+    return array.reduce<Record<K, T>>((ret, item, index) => {
+      const key = indexed ? fn(item, index, array) : fn(item);
       ret[key] = item;
       return ret;
-    }, {});
+    }, {} as Record<K, T>); // eslint-disable-line @typescript-eslint/prefer-reduce-type-parameter
```

Comment:
It seems like this assertion, and disabling of the eslint rule, as a necessary evil, as otherwise the empty object can't be used as the initial value parameter. It's a little redundant to keep the type parameter, but leaving it there for now in anticipation of a better way to handle this issue.

Reply:
you'd have to type the object as Partial<Record<K, T>> otherwise typescript is correct that the types don't match. The empty object doesn't have props for each K

Reply:
Thanks @eranhirsch. Though I think there will have to be some assertions unless I'm mistaken, as if we type the object as `Partial<Record<K, T>>`, then it seems we must also assert the result of the reduce—as it would affect the return type. I've gone with the below for now:

```ts
    const initialvalues: Partial<Record<K, T>> = {}
    return array.reduce((ret, item, index) => {
      const key = indexed ? fn(item, index, array) : fn(item);
      ret[key] = item;
      return ret;
    }, initialvalues) as Record<K, T>;
```

But this is also an option, which I decided against for readability reasons. Though I'm happy to change:

```ts
    return array.reduce<Partial<Record<K, T>>>((ret, item, index) => {
      const key = indexed ? fn(item, index, array) : fn(item);
      ret[key] = item;
      return ret;
    }, {}) as Record<K, T>;
```

Reply:
You are right that we'd have to cast again at the end, that's fine. I would split this into two lines so the casting is more obvious. Assign the reduce into a local variable and then return the result casted to the right type

Reply:
I've just made that change 👍🏾